core/src/main/scala/kafka/log/UnifiedLog.scala

```diff
@@ -1337,13 +1337,17 @@ class UnifiedLog(@volatile var logStartOffset: Long,
       } else if (targetTimestamp == ListOffsetsRequest.MAX_TIMESTAMP) {
         // Cache to avoid race conditions. `toBuffer` is faster than most alternatives and provides
         // constant time access while being safe to use with concurrent collections unlike `toArray`.
-        val segmentsCopy = logSegments.asScala.toBuffer
-        val latestTimestampSegment = segmentsCopy.maxBy(_.maxTimestampSoFar)
-        val latestTimestampAndOffset = latestTimestampSegment.readMaxTimestampAndOffsetSoFar
-
-        Some(new TimestampAndOffset(latestTimestampAndOffset.timestamp,
-          latestTimestampAndOffset.offset,
-          latestEpochAsOptional(leaderEpochCache)))
+        val latestTimestampSegment = logSegments.asScala.toBuffer.maxBy[Long](_.maxTimestampSoFar)
+        // cache the timestamp and offset
+        val maxTimestampSoFar = latestTimestampSegment.readMaxTimestampAndOffsetSoFar
+        // lookup the position of batch to avoid extra I/O
+        val position = latestTimestampSegment.offsetIndex.lookup(maxTimestampSoFar.offset)
+        val lpc = latestEpochAsOptional(leaderEpochCache)
+        Some(latestTimestampSegment.log.batchesFrom(position.position).asScala
+          .find(_.maxTimestamp() == maxTimestampSoFar.timestamp)
+          .flatMap(batch => batch.offsetOfMaxTimestamp().asScala.map(new TimestampAndOffset(batch.maxTimestamp(), _, lpc)))
+          // return the base offset for backward compatibility if there is no batches
+          .getOrElse(new TimestampAndOffset(RecordBatch.NO_TIMESTAMP, latestTimestampSegment.baseOffset(), lpc)))
```

Comment:
Hmm, still not sure about this. If we can't find the maxTimestamp, intuitively, it seems that we should return -1 for both timestamp and offset?

Reply:
I am ok to return -1. However, it seems that we return base offset before when we do not find the max timestamp (no batch exists). Hence, the main reason of returning base offset is backward compatibility.

Reply:
If it is fine to change the behavior, we can just retune None to build the response with unknown offset/timestamp.

https://github.com/apache/kafka/blob/ee61bb721eecb0404929f125fe43392f3d024453/core/src/main/scala/kafka/server/KafkaApis.scala#L1146

Reply:
Got it. If there is no timestamp index, we initialize `maxTimestampAndOffsetSoFar` to `TimestampOffset(RecordBatch.NO_TIMESTAMP, baseOffset())`. That's why it picks up the base offset. However, it doesn't seem intuitive for the user. Returning None seems better.

Reply:
sure. https://github.com/apache/kafka/pull/15621/commits/8a7ed30692bd070fb4160a6cbc76a868484529c3 return none and add related test